### PR TITLE
Issue #24534: Msg for wrong package Resource annotation

### PR DESCRIPTION
**Pull request by tkburroughs**
Log an audit message when a Jakarta EE 9+ application includes use of javax.annotation.Resource, javax.annotation.PostConstruct, or javax.annotation.PreDestroy

for #24534

### CODE REVIEW

**Comment by tkburroughs** [comment]
#build

**Review by tbitonti — DISMISSED** [review]

**Comment by posmith** [comment]
L2 message review completed

File: dev/com.ibm.ws.container.service/resources/com/ibm/ws/container/service/resources/Messages.nlsprops [hunk 1]
```diff
@@ -166,4 +166,6 @@ ordered.list.already.contains.CWWKM0482E=CWWKM0482E: WebSphere Application Serve
 ordered.list.already.contains.CWWKM0482E.explanation=CMSG0001
 ordered.list.already.contains.CWWKM0482E.useraction=See the problem determination information on the WebSphere Application Server Support web page: http://www.ibm.com/software/webservers/appserv/was/support.
 
-
+wrong.annotation.package.CWWKM0483I=CWWKM0483I: The {0} annotation will be ignored on classes in the {1} module. Jakarta EE applications must use the {2} annotation. The affected classes are: {3}
```
Comment (helyarp) [inline dev/com.ibm.ws.container.service/resources/com/ibm/ws/container/service/resources/Messages.nlsprops]: We are not supposed to use future tense in our messages. Please change "will be" to "is" unless this message requires "will be" to be technically correct.

File: dev/com.ibm.ws.container.service/resources/com/ibm/ws/container/service/resources/Messages.nlsprops [hunk 2]
```diff
@@ -166,4 +166,6 @@ ordered.list.already.contains.CWWKM0482E=CWWKM0482E: WebSphere Application Serve
...
+wrong.annotation.package.CWWKM0483I.explanation=The application contains a Java EE annotation with the 'javax' package name, but the application server has Jakarta EE features enabled. The application server will not process the 'javax' package name annotation.
```
Comment (helyarp) [inline dev/com.ibm.ws.container.service/resources/com/ibm/ws/container/service/resources/Messages.nlsprops]: Please change "will" to "did" unless "will" is required to make the message technically correct.

**Review by helyarp — COMMENTED** [review]
Suggested changes

File: dev/com.ibm.ws.container.service/resources/com/ibm/ws/container/service/resources/Messages.nlsprops (hunk 1 above)
Comment (tkburroughs) [reply dev/com.ibm.ws.container.service/resources/com/ibm/ws/container/service/resources/Messages.nlsprops]: done

File: dev/com.ibm.ws.container.service/resources/com/ibm/ws/container/service/resources/Messages.nlsprops (hunk 2 above)
Comment (tkburroughs) [reply dev/com.ibm.ws.container.service/resources/com/ibm/ws/container/service/resources/Messages.nlsprops]: done

**Comment by tkburroughs** [comment]
@helyarp changes made as suggested, thanks.

**Comment by tkburroughs** [comment]
#libby
#build

**Comment by tkburroughs** [comment]
#build

**Comment by tkburroughs** [comment]
Latest build with message improvements was good; ready to merge.

**Review by helyarp — COMMENTED** [review]
Looks good!

File: dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/AnnotationsImpl.java [hunk 3]
```diff
@@ -735,7 +744,76 @@ protected void addExternalToClassSource() {
...
+    @FFDCIgnore(ClassNotFoundException.class)
+    private boolean jakartaCommonAnnotationAPIs() {
+        if (classLoader == null) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "Jakarta Common Annotations not present - ClassLoader not set");
+            }
+            return false;
+        }
+        try {
+            classLoader.loadClass("jakarta.annotation.Resource");
```
Comment (olendvcook) [inline dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/AnnotationsImpl.java]: Could be replaced with JAKARTA_RESOURCE variable

File: dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/AnnotationsImpl.java [hunk 4]
```diff
@@ -735,7 +744,76 @@ protected void addExternalToClassSource() {
...
+     *            OR of scan policy values.
+     */
+    private void checkForWrongPackageCommonAnnotations(AnnotationTargets_Targets targets, int scanPolicies) {
+        if (modName == null || !jakartaCommonAnnotationAPIs()) {
+            return;
+        }
+
+        Set<String> foundClasses = new TreeSet<String>(); // provides natural ordering
+
+        // Look for @Resoruce at class, field, and method level and nested in @Resoruces at class level
```
Comment (olendvcook) [inline dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/AnnotationsImpl.java]: "@Resoruce" - spelled wrong in the comment

File: dev/io.openliberty.injection.jakarta_fat/fat/src/io/openliberty/injection/jakarta/fat/tests/WrongPackageTest.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+     */
+    @Test
+    public void testWrongPackageCommonAnnotations_EE9() throws Exception {
+        try {
+            ee9server.startServer();
+            assertEquals("Expected CWWKM0483I message not found for @Resource in EJB module : " + CWWKM0483I_EJB_MOD_RESOURCE, 1,
+                         ee9server.findStringsInLogsUsingMark(CWWKM0483I_EJB_MOD_RESOURCE, ee9server.getDefaultLogFile()).size());
+            assertEquals("Expected CWWKM0483I message not found for @PreDestroy in EJB module : " + CWWKM0483I_EJB_MOD_PREDESTROY, 1,
+                         ee9server.findStringsInLogsUsingMark(CWWKM0483I_EJB_MOD_PREDESTROY, ee9server.getDefaultLogFile()).size());
+            assertEquals("Expected CWWKM0483I message not found for @Resoruce in WEB module : " + CWWKM0483I_WEB_MOD_RESOURCE, 1,
```
Comment (olendvcook) [inline dev/io.openliberty.injection.jakarta_fat/fat/src/io/openliberty/injection/jakarta/fat/tests/WrongPackageTest.java]: @Resoruce again

File: dev/io.openliberty.injection.jakarta_fat/fat/src/io/openliberty/injection/jakarta/fat/tests/WrongPackageTest.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    @Test
+    @MinimumJavaLevel(javaLevel = 11)
+    public void testWrongPackageCommonAnnotations_EE10() throws Exception {
+        try {
+            ee10server.startServer();
+            assertEquals("Expected CWWKM0483I message not found for @Resource in EJB module : " + CWWKM0483I_EJB_MOD_RESOURCE, 1,
+                         ee10server.findStringsInLogsUsingMark(CWWKM0483I_EJB_MOD_RESOURCE, ee10server.getDefaultLogFile()).size());
+            assertEquals("Expected CWWKM0483I message not found for @PreDestroy in EJB module : " + CWWKM0483I_EJB_MOD_PREDESTROY, 1,
+                         ee10server.findStringsInLogsUsingMark(CWWKM0483I_EJB_MOD_PREDESTROY, ee10server.getDefaultLogFile()).size());
+            assertEquals("Expected CWWKM0483I message not found for @Resoruce in WEB module : " + CWWKM0483I_WEB_MOD_RESOURCE, 1,
```
Comment (olendvcook) [inline dev/io.openliberty.injection.jakarta_fat/fat/src/io/openliberty/injection/jakarta/fat/tests/WrongPackageTest.java]: @Resoruce again

File: dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/WrongPackageServlet.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    @EJB
+    JakartaStatelessBean bean7;
+
+    /**
+     * This test verifies a warning is logged when the javax.annotation.Resource annotation
+     * is used with Jakarta EE features and the annotation is ignored. There should be one
+     * warning per module. The jakarta.annotation.Resource annotation works as expected.
+     */
+    public void testWrongPackageCommonAnnotations() throws Exception {
+        assertNotNull("jakarta Resrouce UserTransaction is null", userTx);
```
Comment (olendvcook) [inline dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/WrongPackageServlet.java]: Resrouce

**Review by olendvcook — DISMISSED** [review]

**Comment by olendvcook** [comment]
~~I'm finding it hard to google if Jakarta @Resources is a valid annotation, or if they scrapped it, but I guess I'm assuming based on these changes that it doesn't exist, since this seems to only mention javax @Resources.~~

Nevermind, Jakarta @Resources does exist but the code here is grouping it up with the jakarta Resource message.

File: dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/AnnotationsImpl.java (hunk 3 above)
Comment (tkburroughs) [reply dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/AnnotationsImpl.java]: changed

File: dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/AnnotationsImpl.java (hunk 4 above)
Comment (tkburroughs) [reply dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/AnnotationsImpl.java]: fixed

File: dev/io.openliberty.injection.jakarta_fat/fat/src/io/openliberty/injection/jakarta/fat/tests/WrongPackageTest.java (hunk 5 above)
Comment (tkburroughs) [reply dev/io.openliberty.injection.jakarta_fat/fat/src/io/openliberty/injection/jakarta/fat/tests/WrongPackageTest.java]: fixed

File: dev/io.openliberty.injection.jakarta_fat/fat/src/io/openliberty/injection/jakarta/fat/tests/WrongPackageTest.java (hunk 6 above)
Comment (tkburroughs) [reply dev/io.openliberty.injection.jakarta_fat/fat/src/io/openliberty/injection/jakarta/fat/tests/WrongPackageTest.java]: fixed

File: dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/WrongPackageServlet.java (hunk 7 above)
Comment (tkburroughs) [reply dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/WrongPackageServlet.java]: fixed

**Comment by tkburroughs** [comment]
#libby
#build

**Comment by tkburroughs** [comment]
@helyarp At design review it was requested I update the message to include the application name, so I've updated to add that.  Please review message gain... Thanks.

**Comment by tkburroughs** [comment]
#libby
#build

**Comment by tkburroughs** [comment]
@olendvcook per design review, I've
- updated message to include the application name (and updated code)
- updated the test to check for correct application name in msg
- updated the test to include standalone web application (war) to ensure proper application name

Please review again... thanks.

**Review by helyarp — COMMENTED** [review]
Message changes look good!

File: dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageEJB.jar/src/io/openliberty/injection/jakarta/ejb/JakartaSingletonResourcesBean.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    @Resource
+    EJBContext ctx;
+
+    public void verifyInjection() {
+        assertNotNull("jakarta Resource EJBContext is null", ctx);
+
+        try {
+            Context namingCtx = new InitialContext();
+            SessionContext sctx = (SessionContext) namingCtx.lookup("java:comp/env/SessionContext");
+            fail("lookup of javax Resrouce was successful : " + sctx);
```
Comment (olendvcook) [inline dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageEJB.jar/src/io/openliberty/injection/jakarta/ejb/JakartaSingletonResourcesBean.java]: Resrouce

File: dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageEJB.jar/src/io/openliberty/injection/jakarta/ejb/JakartaStatelessBean.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    EJBContext ctx;
+
+    @javax.annotation.Resource
+    SessionContext sctx;
+
+    @EJB
+    JakartaSingletonBean bean;
+
+    public void verifyInjection() {
+        assertNotNull("jakarta Resrouce EJBContext is null", ctx);
```
Comment (olendvcook) [inline dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageEJB.jar/src/io/openliberty/injection/jakarta/ejb/JakartaStatelessBean.java]: Resrouce

File: dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/JakartaSingletonWarBean.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    EJBContext ctx;
+
+    @javax.annotation.Resource
+    SessionContext sctx;
+
+    @EJB
+    JakartaStatelessWarBean bean;
+
+    public void verifyInjection() {
+        assertNotNull("jakarta Resrouce EJBContext is null", ctx);
```
Comment (olendvcook) [inline dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/JakartaSingletonWarBean.java]: Resrouce

File: dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/JakartaStatelessWarBean.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    EJBContext ctx;
+
+    @javax.annotation.Resource
+    SessionContext sctx;
+
+    @EJB
+    JakartaSingletonWarBean bean;
+
+    public void verifyInjection() {
+        assertNotNull("Jakarta Resrouce EJBContext is null", ctx);
```
Comment (olendvcook) [inline dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/JakartaStatelessWarBean.java]: Resrouce

**Review by olendvcook — DISMISSED** [review]

**Comment by olendvcook** [comment]
I found some more misspellings that I missed the first review.

File: dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageEJB.jar/src/io/openliberty/injection/jakarta/ejb/JakartaSingletonResourcesBean.java (hunk 8 above)
Comment (tkburroughs) [reply dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageEJB.jar/src/io/openliberty/injection/jakarta/ejb/JakartaSingletonResourcesBean.java]: fixed

File: dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/JakartaSingletonWarBean.java (hunk 10 above)
Comment (tkburroughs) [reply dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/JakartaSingletonWarBean.java]: fixed

File: dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/JakartaStatelessWarBean.java (hunk 11 above)
Comment (tkburroughs) [reply dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/JakartaStatelessWarBean.java]: fixed

File: dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageEJB.jar/src/io/openliberty/injection/jakarta/ejb/JakartaStatelessBean.java (hunk 9 above)
Comment (tkburroughs) [reply dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageEJB.jar/src/io/openliberty/injection/jakarta/ejb/JakartaStatelessBean.java]: fixed

**Comment by tkburroughs** [comment]
@olendvcook Fixed.

**Comment by tkburroughs** [comment]
#libby
#build

**Review by olendvcook — APPROVED** [review]

**Comment by tkburroughs** [comment]
All test failures in build linked to existing know defects; ready to merge.

**Pull request merged**